### ISSUE TITLE
handler does not work on PHP7+

### ISSUE THREAD

**Issue by marcing** [1ma/RedisSessionHandler#3]
Error #2 session_regenerate_id(): Session object destruction failed.  ID: user (path: tcp://127.0.0.1:6379?database=2&timeout=2)

Steps to reproduce:
enter website
flush redis:
redis-cli> SELECT 2
redis-cli> FLUSHDB
reload website

**Comment by 1ma** [comment]
You mean [this](https://github.com/1ma/RedisSessionHandler/blob/master/tests/webroot/visit-counter.php) website? I can't reproduce (using PHP 7.0.15 and Redis 3.0.7).

Do you think that you could provide me a sample php script like the one I linked and the sequence of curl and redis commands that reproduce the issue? I would integrate that into the testing suite.

**Comment by marcing** [comment]
Yes, but I forgot to mention that it is after you have a session cookie already and delete entries from redis, so when you simulate session fixation, exact error message is as follows:

Warning: session_regenerate_id(): Cannot call session save handler in a recursive manner in RedisSessionHandler.php on line 152

Warning: session_regenerate_id(): Session object destruction failed. ID: user (path: tcp://127.0.0.1:6379?database=2&timeout=2) in RedisSessionHandler.php on line 152

**Comment by 1ma** [comment]
I'm sorry, I still can't reproduce this. FWIW I added an e2e test to the suite (see commit above) but it passes. It might be an issue related to Windows (I saw that your first warning messages included some Windows path) but this is just speculation.

However there's something that suprised me: neither of the existing releases (v0.9.0 and v0.9.1) have a `session_regenerate_id` call on line 152 of the handler. The first relase had it on L105 and the second one on L115 :thinking:

**Comment by marcing** [comment]
Dear 1ma.

Yes the line number is incorrect, because I added the parameters support to the open() method to be able to test your handler. I do not run sessions on database 0, but use it for cache.

I encountered the issue first on linux debian, but later reproduced it on a windows machine, after which I created this ticket. The error message is very logical when you think about it. You have this code:

`    public function read($session_id)
    {
        if ( $this->mustRegenerate($session_id)) {
            // Regenerating the ID will call destroy(), close(), open(), create_sid() and read() in this order.
            // It will also signal the PHP internals to include the 'Set-Cookie' with the new ID in the HTTP response.
            session_regenerate_id(true);

            return '';
        }`

and you even write in the comment that session_regenerate_id calls read(), so it basically tries to invoke itself: read -> session_regenerate_id -> read and it crashes.
I am sure you can reproduce it on PHP7 (does not happen on PHP5, because in PHP5 "read" is not called). You need enter "if ( $this->mustRegenerate($session_id)) {" condition to recreate it, which will start the recursive calls.

I attach my test script, which I use to display the error message:
https://dl.dropboxusercontent.com/u/11435743/redis.zip

Thanks!
Marcin

**Comment by 1ma** [comment]
Hi Marcin, I can finally reproduce this on PHP 7.1.2. The error is not due to infinite recursion, as `mustRegenerate` can return `true` only once, but rather to a [recent change](https://github.com/php/php-src/pull/2196) introduced in the 7.1.x branch. Apparently calling some  `session_*` functions inside the session handler is considered "an abuse" and is now forbidden.

In the meantime all releases of PHP 7.0 work well, but I expect to fix this in the next 24-48h.

**Comment by marcing** [comment]
Dear 1ma,

Thank you, I forgot to mention that I am using 7.1 on both linux/windows. I somehow thought this also affects 7.0. Either way I was not aware of this change as well. I am happy that you managed to find this information.

I wish you luck and thank you.
Marcin

**Comment by 1ma** [comment]
It now works on my 7.1 testing environment. Would you be so kind to test if this latest version works for you too?

https://raw.githubusercontent.com/1ma/RedisSessionHandler/d5634f57e389ce17db07cd56a8e0c2ead3f70df9/src/RedisSessionHandler.php

If it does I'll tag it and ship a new release.

**Comment by marcing** [comment]
Dear 1ma,

The error is gone, but I noticed some things you need to consider.

1. Before invoking a "setcookie", you should probably set the params taken from:
http://php.net/manual/en/function.session-get-cookie-params.php
or ini_get('session.*) 

The values cookie_path, cookie_domain, cookie_secure, cookie_httponly may differ for every setup.

2. In some cases, where sending cookie header is not desired, you should not send the cookie. There is a flag for this: session.use_cookies.

When this is done, I will try to test it on one of my production setups.

Thank you,
Marcin

**Comment by 1ma** [comment]
1. I didn't know about this function, however in my tests I'm seeing that all these values are already being read from the global configuration if they are not passed explicitly to `setcookie`. No further change seems necessary.

![custom cookie params](http://i.imgur.com/IZjjZk0.png)

2. I'm not sold on this one. If the handler fails to send the cookie the current implementation will be broken, or else it'll also need to support [URL-based sessions](http://www.phpddt.com/usr/uploads/2013/09/2521893358.jpg), and honestly I don't want to do that. Do you actually need this?

**Comment by marcing** [comment]
**setcookie:**
This is how Set-Cookie header looks like when session handler sets it:
Set-CookiePHPSESSID=sp0srg6ab9oa0fardo2qf3duh0; expires=Thu, 23-Feb-2017 09:08:28 GMT; Max-Age=20; path=/redis/; domain=locahost; HttpOnly
This is how Set-Cookie header looks like when you set it:
Set-CookiePHPSESSID=ang9n9f5l452s05mp4qpehp8l2; path=/

Your proof is merely a coincidence, it works like this:
1. session_start(); sets a cookie using session_get_cookie_params [correctly]
2. you set a cookie under the same name, overwriting it's value
If cookie doesn't exist for some reason at the time you regenerate the ID, it will be saved with false parameters. This may lead to rare and hard to track errors.

**use_cookies:**
It is not about URL-based sessions, it is only about adding one condition: if use_cookies = 0, do not invoke setcookie. In our system there are cases where session ID is not passed in a cookie, but passed directly to session_id() in order to read session data. This parameter was added to PHP not without reason :)

Cheers

**Comment by mrols** [comment]
Can we have this update tagged?

We would want our core library to require this library up to this update.

Thanks!

**Comment by marcing** [comment]
The issues are still not fixed, I personally dropped the idea of using this plugin as I have serious doubts if author is capable of making it production ready. Not really sure if anyone should be using it.

**Comment by 1ma** [comment]
@michaerum there you go. Care to share your PHP version/the no. of concurrent sessions you handle/anything really?

@marcing sorry to drop the ball like that, but I see no more issues. As already explained, I am able to customize the Set-Cookie string with nothing more than php.ini settings, no code changes in the handler seem necessary. Here's exactly the same Set-Cookie string from your last message:
![moar custom cookie params](http://i.imgur.com/PLkAoW7.png)

WRT to the `use_cookies` issue, are you aware that if I do what you suggest and then set `use_cookies` to `0` the handler would just not work at all? Unless you explain me like I'm 5 how sessions would work in this scenario I won't do this, period.

And as a general reminder, note that the package follows semantic versioning. Thus releases lower than 1.0 indeed mean "not production ready, probably". Thank you for stepping out to help.

**Comment by 1ma** [comment]
I'm closing this since the original issue is actually fixed.
